src/libp2p_transport_tcp.erl

```diff
@@ -387,66 +387,72 @@ handle_info({handle_identify, Session, {error, Error}}, State=#state{}) ->
+                                                                                                [{tcp_listen_addrs(S), Pid} | Acc];
+                                                                                            false ->
+                                                                                                Acc
+                                                                                        end
+                                                                                end;
+                                                                            false ->
+                                                                                Acc
+                                                                        end
+                                                                end, [], MyListenAddrs)
+                                            end,
+                    %% don't use lists flatten here as it flattens too much
+                    NewListenAddrs = lists:foldl(fun(E, A) -> E ++ A end, [], (element(1, lists:unzip(NewListenAddrsWithPid)))),
+                    case lists:member(LocalAddr, NewListenAddrs) orelse (ListenAddrs == [] andalso NewListenAddrs /= []) of
+                        true ->
+                            %% they have changed, or we never had any; add any new ones and remove any old ones
+                            ObservedAddr = libp2p_identify:observed_addr(Identify),
+                            %% find listen addresses for this transport
+                            MyListenAddrs = [ LA || LA <- ListenAddrs, match_addr(LA, TID) /= false ],
+                            RemovedListenAddrs = MyListenAddrs -- NewListenAddrs,
+                            lager:debug("Listen addresses changed: ~p -> ~p", [MyListenAddrs, NewListenAddrs]),
+                            [ libp2p_config:remove_listener(TID, A) || A <- RemovedListenAddrs ],
+                            %% we can simply re-add all of the addresses again, overrides are fine
+                            [ libp2p_config:insert_listener(TID, LAs, P) || {LAs, P} <- NewListenAddrsWithPid],
+                            PB = libp2p_swarm:peerbook(TID),
+                            libp2p_peerbook:changed_listener(PB),
+                            {noreply, record_observed_addr(RemoteP2PAddr, ObservedAddr, State)};
+                        false ->
+                            lager:debug("identify response with local address ~p that is not a listen addr socket ~p, ignoring",
+                                        [LocalAddr, NewListenAddrs]),
+                            %% this is likely a discovery session we dialed with unique_port
+                            %% we can't trust this for the purposes of the observed address
+                            {noreply, State}
+                    end
             end
+    catch
+        _:_ ->
```

Comment:
should we be logging things here so we can more selectively filter later?

Reply:
or maybe a (case catch) to keep it in the case statements and reduce one indent level in that insanely large code block?

Reply:
I realize there are a couple of places an error can be thrown so the top level try block may be the best way to do it, but this section of code is getting pretty unwieldy, no?